shared/ssz/hash_cache_test.go

```diff
@@ -0,0 +1,213 @@
+package ssz
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+)
+
+func TestHashKeyFn_OK(t *testing.T) {
+	mRoot := &root{
+		Hash: common.HexToHash("0x0123456"),
+	}
+
+	key, err := hashKeyFn(mRoot)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if key != mRoot.Hash.Hex() {
+		t.Errorf("Incorrect hash key: %s, expected %s", key, mRoot.Hash.Hex())
+	}
+}
+
+func TestHashKeyFn_InvalidObj(t *testing.T) {
+	_, err := hashKeyFn("bad")
+	if err != ErrNotMerkleRoot {
+		t.Errorf("Expected error %v, got %v", ErrNotMerkleRoot, err)
+	}
+}
+
+func TestObjCache_byHash(t *testing.T) {
+	cache := newHashCache()
+
+	byteSl := [][]byte{{0, 0}, {1, 1}}
+	mr, err := merkleHash(byteSl)
+	if err != nil {
+		t.Fatal(err)
+	}
+	hs, err := hashedEncoding(reflect.ValueOf(byteSl))
+	if err != nil {
+		t.Fatal(err)
+	}
+	exists, _, err := cache.RootByEncodedHash(bytesutil.ToBytes32(hs))
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	if exists {
+		t.Error("Expected block info not to exist in empty cache")
+	}
+
+	if _, err := cache.MerkleHashCached(byteSl); err != nil {
+		t.Fatal(err)
+	}
+
+	exists, fetchedInfo, err := cache.RootByEncodedHash(bytesutil.ToBytes32(hs))
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !exists {
+		t.Error("Expected blockInfo to exist")
+	}
+	if bytes.Compare(mr, fetchedInfo.MarkleRoot) != 0 {
+		t.Errorf(
+			"Expected fetched info number to be %v, got %v",
+			mr,
+			fetchedInfo.MarkleRoot,
+		)
+	}
+	if fetchedInfo.Hash != bytesutil.ToBytes32(hs) {
+		t.Errorf(
+			"Expected fetched info hash to be %v, got %v",
+			hs,
+			fetchedInfo.Hash,
+		)
+	}
+}
+
+func TestMerkleHashWithCache(t *testing.T) {
+	cache := newHashCache()
+	for i := 0; i < 200; i++ {
+
+		runMerkleHashTests(t, func(val [][]byte) ([]byte, error) {
+			return merkleHash(val)
+		})
+
+	}
+
+	for i := 0; i < 200; i++ {
+
+		runMerkleHashTests(t, func(val [][]byte) ([]byte, error) {
+			return cache.MerkleHashCached(val)
+		})
+
+	}
+
+}
+
+type JunkObject struct {
```

Comment:
Please put this struct and helper functions at the top of the test file

Reply:
done